aws/rust-runtime/aws-inlineable/src/body_with_checksum.rs

```diff
@@ -0,0 +1,195 @@
+    let aws_chunked_body_options = aws_http::content_encoding::AwsChunkedBodyOptions::new()
+        .with_stream_length(original_body_size)
+        .with_trailer_len(body.trailer_length());
+
+    let body = aws_http::content_encoding::AwsChunkedBody::new(body, aws_chunked_body_options);
+    let encoded_content_length = body.size_hint().exact().ok_or_else(|| {
+        aws_smithy_http::operation::BuildError::Other(Box::new(
+            ChecksumValidatedRequestError::UnsizedRequestBody,
+        ))
+    })?;
+    let headers = request.headers_mut();
+    headers.insert(
+        http::header::CONTENT_LENGTH,
+        http::HeaderValue::from(encoded_content_length),
+    );
+    headers.insert(
+        http::header::HeaderName::from_static("x-amz-decoded-content-length"),
+        http::HeaderValue::from(original_body_size),
+    );
+    headers.insert(
+        http::header::HeaderName::from_static("x-amz-trailer"),
+        checksum_trailer_name.into(),
+    );
+    headers.insert(
+        http::header::CONTENT_ENCODING,
+        http::HeaderValue::from_str(aws_http::content_encoding::header_value::AWS_CHUNKED)
+            .map_err(|err| aws_smithy_http::operation::BuildError::Other(Box::new(err)))?,
+    );
+
+    let mut body =
+        aws_smithy_http::body::SdkBody::from_dyn(http_body::combinators::BoxBody::new(body));
+
+    std::mem::swap(request.body_mut(), &mut body);
+
+    Ok(())
+}
+
+/// Given a `Response<SdkBody>`, checksum algorithm name, and pre-calculated checksum, return a
+/// `Response<SdkBody>` where the body will processed with the checksum algorithm and checked
+/// against the pre-calculated checksum.
+pub fn build_checksum_validated_sdk_body(
+    body: aws_smithy_http::body::SdkBody,
+    checksum_algorithm: &str,
+    precalculated_checksum: bytes::Bytes,
+) -> aws_smithy_http::body::SdkBody {
+    tracing::trace!(
+        "wrapping response body in {} checksum validator, expected checksum is {:#?}",
+        checksum_algorithm,
+        precalculated_checksum,
+    );
+    let body = aws_smithy_checksums::body::ChecksumValidatedBody::new(
+        body,
+        checksum_algorithm,
+        precalculated_checksum.clone(),
+    );
+    aws_smithy_http::body::SdkBody::from_dyn(http_body::combinators::BoxBody::new(body))
+}
+
+/// Given the name of a checksum algorithm and a `HeaderMap`, extract the checksum value from the
+/// corresponding header as `Some(Bytes)`. If the header is unset, return `None`.
+pub fn check_headers_for_precalculated_checksum(
+    headers: &http::HeaderMap<http::HeaderValue>,
+    response_algorithms: &[&str],
+) -> Option<(&'static str, bytes::Bytes)> {
+    // TODO sort response_algorithms in priority order based on `aws_smithy_checksums::CHECKSUM_HEADERS_IN_PRIORITY_ORDER`
+    let response_algorithms = response_algorithms.iter().map(|&algorithm| {
+        aws_smithy_checksums::checksum_algorithm_to_checksum_header_name(algorithm)
```

Comment:
nit: naming, include checksum in the module name so you don't need to repeat it twice in the name

```checksums::algorithm2header``` or something